### PR TITLE
metrics: basic infra based on gRPC interceptors

### PR DESCRIPTION
**Pull request by uael**

### CODE REVIEW

File: avatar/pandora_client.py [hunk 1]
```diff
@@ -169,9 +173,20 @@ def security_storage(self) -> security_grpc_aio.SecurityStorage:
     @property
     def aio(self) -> 'PandoraClient.Aio':
         if not self._aio:
-            self._aio = PandoraClient.Aio(grpc.aio.insecure_channel(self.grpc_target))
+            self._aio = PandoraClient.Aio(
+                grpc.aio.insecure_channel(self.grpc_target, interceptors=aio_interceptors(self))
+            )
         return self._aio
 
+    @property
```
Comment (uael) [inline avatar/pandora_client.py:181]: Todo: remove next two properties

File: avatar/metrics/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+## Perfetto traces
+
+For convenience, `trace_pb2.py` and `trace_pb2.pyi` are pre-generated.
+
+To regenerate them run the following:
+
+```
+pip install protoc-exe
+protoc trace.proto --pyi_out=./ --python_out=./
+```
```
Comment (uael) [inline avatar/metrics/README.md:14]: Add a comment about removing `Union[T, _Mapping]` from generated `.pyi`

File: avatar/metrics/trace.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+                track_uuid=self.callsite.device.uuid,
+                debug_annotations=None
+                if self.message is None
+                else [DebugAnnotation(string_value=message_prettifier(f"{self.message}"))],
+            ),
+            trusted_packet_sequence_id=process_id,
+        )
+
+    def stringify(self, direction: str) -> str:
+        message_pretty = message_prettifier(f"{self.message}")
```
Comment (uael) [inline avatar/metrics/trace.py:194]: Empty string `if self.message is None` (fix for `Non`)

File: avatar/metrics/trace.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+packets: List[TracePacket] = []
+genesis: int = time.monotonic_ns()
+uuid: int = 0
+
+
+def process_name(test: BaseTestClass) -> str:
+    return f"{test.__class__.__name__}.{test.current_test_info.name}"
+
+
+def next_uuid() -> int:
```
Comment (DeltaEvo) [inline avatar/metrics/trace.py:51]: nit: `next_id`, a `uuid` is a different thing

File: avatar/metrics/trace.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    from avatar import PandoraDevices
+    from avatar.pandora_client import PandoraClient
+else:
+    PandoraClient = object
+    PandoraDevices = object
+
+
+packets: List[TracePacket] = []
+genesis: int = time.monotonic_ns()
+uuid: int = 0
```
Comment (DeltaEvo) [inline avatar/metrics/trace.py:44]: nit: move that above `next_uuid`

File: avatar/metrics/trace.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+            for packet in packets:
+                f.write(f"{packet}")
+                f.write("----------\n")
+
+        original_teardown_class(self)
+
+    def setup_test(self: BaseTestClass) -> None:
+        global genesis
+        genesis = time.monotonic_ns()
+        assert hasattr(self, "devices")
```
Comment (DeltaEvo) [inline avatar/metrics/trace.py:77]: We should not rely on the test writer to call it's variable "devices", can we add it as argument of `hook_test`

File: avatar/metrics/trace.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        process_id = next_uuid()
+        devices: PandoraDevices = getattr(self, "devices", [])  # type: ignore
+        packets.append(
+            TracePacket(
+                track_descriptor=TrackDescriptor(
+                    uuid=process_id,
+                    process=ProcessDescriptor(pid=process_id, process_name=process_name(self)),
```
Comment (DeltaEvo) [inline avatar/metrics/trace.py:84]: `process_name` only seems to be used here an is a one line function can we inline it ?

File: avatar/metrics/trace.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+    def __init__(self, device: PandoraClient, name: Union[bytes, str], message: Any) -> None:
+        self.at = time.monotonic_ns() - genesis
+        self.name = name if isinstance(name, str) else name.decode('utf-8')
+        self.device = device
+        self.message = message
+        self.events: List[CallEvent] = []
+        self.id = Callsite.next_id()
+
+        device.log.info(f"{self}")
```
Comment (DeltaEvo) [inline avatar/metrics/trace.py:126]: nit: Remove ?

File: avatar/metrics/trace.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        return f"%{self.id} {name_pretty}({message_pretty})"
+
+    def output(self, message: Any) -> None:
+        self.events.append(CallOutput(self, message))
+
+    def input(self, message: Any) -> None:
+        self.events.append(CallInput(self, message))
+
+    def end(self, message: Any) -> None:
+        global packets
```
Comment (DeltaEvo) [inline avatar/metrics/trace.py:142]: nit: Can we take packets as argument ?

File: avatar/metrics/trace.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+    def repl(addr: re.Match[str]) -> str:
+        m = CUnescape(addr.groups()[0])
+        if len(m) > 8:
+            return ":'...'"
+        if len(m) == 4 and m[0] == 0 and m[1] == 0:
+            m = m[2:]
+        return ":'" + ':'.join([f'{x:02X}' for x in m]) + "'"
+
+    return re.sub(r":\"([^\"]+)\"", repl, msg)[:-1]
```
Comment (DeltaEvo) [inline avatar/metrics/trace.py:241]: This is kinda ugly could we find a better way to do this ?

File: cases/config.yml [hunk 11]
```diff
@@ -10,6 +10,6 @@ TestBeds:
 - Name: bumble.bumbles
   Controllers:
     BumbleDevice:
-    - transport: 'tcp-client:127.0.0.1:7300'
-    - transport: 'tcp-client:127.0.0.1:7300'
-    - transport: 'tcp-client:127.0.0.1:7300'
+    - transport: 'tcp-client:127.0.0.1:6402'
+    - transport: 'tcp-client:127.0.0.1:6402'
+    - transport: 'tcp-client:127.0.0.1:6402'
```
Comment (DeltaEvo) [inline cases/config.yml:15]: Is this change intended ?

File: avatar/metrics/trace.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+                track_descriptor=TrackDescriptor(
+                    uuid=process_id,
+                    process=ProcessDescriptor(pid=process_id, process_name=process_name(self)),
+                )
+            )
+        )
+
+        for device in devices:
+            device.process_id = process_id
+            device.uuid = next_uuid()
```
Comment (DeltaEvo) [inline avatar/metrics/trace.py:91]: Could we create those in the `PandoraClient`'s constructor ?

**Review by DeltaEvo — CHANGES_REQUESTED** [review]

File: avatar/metrics/trace.py (hunk 4 above)
Comment (SilverBzH) [reply avatar/metrics/trace.py:51]: Well yeah I agree, just because they were unique in some point, I'll change it

File: avatar/metrics/trace.py (hunk 5 above)
Comment (SilverBzH) [reply avatar/metrics/trace.py:44]: I'd like to keep them above every function this is why they are there. Feel weird if they are under `def process_name`

File: avatar/metrics/trace.py (hunk 6 above)
Comment (SilverBzH) [reply avatar/metrics/trace.py:77]: Agree

File: avatar/metrics/trace.py (hunk 12 above)
Comment (SilverBzH) [reply avatar/metrics/trace.py]: the `device.uuid` I guess so but the process_id need to be created for each test so it is better if it is handled in the trace.py class. Also the `device.uuid` and the `process_id` need to be different so It is easier if we manipulate them at the same place. I'd like to keep it that way unless you see a better way to do it.

File: avatar/metrics/trace.py (hunk 8 above)
Comment (uael) [reply avatar/metrics/trace.py]: Nope, it's the new logging

File: cases/config.yml (hunk 11 above)
Comment (SilverBzH) [reply cases/config.yml:15]: Nope

File: avatar/metrics/trace.py (hunk 10 above)
Comment (SilverBzH) [reply avatar/metrics/trace.py:234]: @uael

File: avatar/metrics/trace.py (hunk 12 above)
Comment (uael) [reply avatar/metrics/trace.py]: I think it's fine for now, unless we create a map of `device` -> `uuid` in `trace.py` instead, @DeltaEvo ?

File: avatar/metrics/trace.py (hunk 10 above)
Comment (uael) [reply avatar/metrics/trace.py:234]: I think it's fine for now, it's only used to logging, a next step would be to recursively generate the debug annotations from the message, but it can be done later (or not at all)

File: avatar/metrics/trace.py (hunk 9 above)
Comment (uael) [reply avatar/metrics/trace.py:135]: It's a bit complex, I prefer like this

File: avatar/metrics/trace.py (hunk 12 above)
Comment (DeltaEvo) [reply avatar/metrics/trace.py]: > I think it's fine for now, unless we create a map of device -> uuid in trace.py instead, @DeltaEvo ?

Yeah I wanted to propose that as well and use a weak map.

> the process_id need to be different so It is easier if we manipulate them at the same place. I'd like to keep it that way unless you see a better way to do it.

If the `process_id` is per test, why are we storing it inside the device ? 

Knowing that I would prefer that we store the `process_id` elsewhere and that we create the `uuid/id` in the PandoraClient

File: avatar/metrics/trace.py (hunk 10 above)
Comment (DeltaEvo) [reply avatar/metrics/trace.py:236]: I prefer that we either remove it or fix it. By keeping it like that, there is a high chance that it'll stay like that.

File: avatar/metrics/trace.py (hunk 12 above)
Comment (uael) [reply avatar/metrics/trace.py]: Let's go for the map then

**Review by SilverBzH — APPROVED** [review]

File: avatar/metrics/trace.py (hunk 10 above)
Comment (uael) [reply avatar/metrics/trace.py]: Done!

File: cases/config.yml (hunk 11 above)
Comment (DeltaEvo) [reply cases/config.yml:15]: It's still there

**Review by DeltaEvo — APPROVED** [review]
